nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp

```diff
@@ -333,6 +336,7 @@ void BtActionServer<ActionT>::populateErrorCode(
   typename std::shared_ptr<typename ActionT::Result> result)
 {
   int highest_priority_error_code = std::numeric_limits<int>::max();
+  std::string highest_priority_error_msg = "";
```

Comment:
Where is this populated for `result->error_msg` after?

Reply:
I don't understand the wording of your question (?after?).  I also don't understand what you mean for me to change it to.

Its likely I will when I try to actually get this to work.

Reply:
We have the `error_msg` sent over the interfaces to the BT nodes. In the BT nodes, we set the outputs to the port. Now, we need a way for the navigators to actually get those values from the blackboard and populate them for the NavigateToPose / NavigateThroughPoses / etc's `error_msg` field. Right now, you just send an empty string and the error messages from the BT nodes are never read from the tree's blackboard like they are for the error codes. 

We should do a similar pattern as the error codes:
- BT Nodes have a default output port for the error code key 
- The BT navigator has a parameter for the list of error code keys to search
- In this part of the code, we use the value of that parameter to search for the error codes and use the highest parameter value

This lets us handle a case where multiple things may be in error but to return us the highest priority one (though functionally that's unlikely ATM). So, we should have similarly separated error msg id values for each of the BT action nodes (and update the defualt BT XMLs in Nav2 to use them) and set a parameter with the list of them in the BT navigator and copy this logic for the error messages.

Perhaps adjust the error code id parameter to be a map instead: `error_code_id: error_msg_id` so that we  can correlate the highest value error code ID with its associated message.

Reply:
> Perhaps adjust the error code id parameter to be a map instead: error_code_id: error_msg_id 
> so that we can correlate the highest value error code ID with its associated message.

By this do you maybe mean change the type of the data stored in the behavior tree port from an int to a complex type?  like this - https://www.behaviortree.dev/docs/tutorial-basics/tutorial_03_generic_ports#parsing-a-string ?

Reply:
Oh no, we should change the `error_code_names` param from a list of error code blackboard keys to something else that correlates the codes with the messages you're adding 

For example: having key value pairs (which is perhaps a little unnatural) or for each of the names we have a param `error_msg_name` which contains its specific error message blackboard key. Then, when we find the highest priority code, we know which message key it belongs to 

https://docs.nav2.org/configuration/packages/configuring-bt-navigator.html

Reply:
I know this is a bit subtle, but if you search for `error_code_names` and trace that, it should be clear to you
- We have a param `error_code_names` that we get with a list of blackboard keys of the action server's error IDs. 
- These are correlated with the error code IDs in the BT XMLs. Its essentially a list of all the BT XML error IDs we need to look at for error codes. We need this param since a user can execute arbitrary BTs with arbitrary error code IDs so they need to tell us all the things we should look for
- At error, we look at them and take the highest valued error code to send back in the Navigator's result, which you found in `populateErrorCode`

Reply:
I think I get it, we need to namespace both the `error_code` and its `error_msg` on the BT blackboard so that the highest priority out of the targets we care about can be selected.  It does feels like alot of duplication in the xml will be required.  IOW for each error_code_id attribute in the xml there will now also be an error_msg attribute.  

```
i.e 
<ComputePathToPose goal="{updated_goal}" path="{path}" planner_id="{selected_planner}" error_code_id="{comp
ute_path_error_code} error_msg="{compute_path_error_msg}"/>
```
ASIDE: The loop variable name `error_code` over `error_code_names_` in populateErrorCode confused me for too long.  I now understand that it allows remapping of names of ports. https://www.behaviortree.dev/docs/tutorial-basics/tutorial_02_basic_ports

Ok on with making sure I have what you want correct.
Currently the nav2_param.yaml file looks like this
```
bt_navigator:
  ros__parameters:
    use_sim_time: true
    ...
    error_code_names:
      - compute_path_error_code
      - follow_path_error_code
      # - smoother_error_code, navigate_to_pose_error_code, navigate_through_poses_error_code, etc
 ```  
 
I think you want 1.)
```
    error_code_names:
      - compute_path_error_code: compute_path_error_msg
      - follow_path_error_code: follow_path_error_msg
      - my_bt_error_code: my_wildly_named_bt_error_msg_because_i_can
```

* Do you want an automatic fallback, with a warning?  i.e replace `_code$` with `_msg$`, if the msg parameter is not present?
* or Should parsing the error_code_names parameter just throw an error to force the param file to be updated?
* If during a populateErrorCode operation, a behavior tree does not have the corresponding `error_msg` attribute should I set the result->error_msg to say something like "'error_msg' port not found in behaviour tree" and log an error?


OR is a whole different construct preferable 2.)
```
    error_code_name_map:
      - code: compute_path_error_code
        msg: compute_path_error_msg
      - code: follow_path_error_code
        msg: follow_path_error_msg
      - code: my_bt_error_code
        msg: my_wildly_named_bt_error_msg_because_i_can
```

Reply:
Looking deeper the ROS2 parameter system does not support full blown yaml.
So I cannot see how the above options would work.

Maybe it could be 
``` 
    error_code_name_prefixes:
       - compute_path
       - follow_path
       - my_bt_action
```

and `_error_code` and `_error_msg` are automatically added as suffixes.

Reply:
I like that!